src/websocket/index.ts

```diff
@@ -0,0 +1,66 @@
+/**
+ * Copyright (c) Hathor Labs and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import BaseWebSocket, { WsOptions } from './base';
+
+/**
+ * Handles websocket connections and message transmission
+ *
+ * @class
+ * @name WS
+ */
+class WalletWebSocket extends BaseWebSocket {
+  constructor(options: WsOptions) {
+    super(options);
+  }
+
+  /**
+   * Handle message receiving from websocket
+   *
+   * @param {Object} evt Event that has data (evt.data) sent in the websocket
+   */
+  onMessage(evt) {
+    const message = JSON.parse(evt.data)
+    const _type = message.type.split(':')[0]
+    if (_type === 'pong') {
+      this.onPong();
+    } else {
+      // The websoket might be exchanging many messages and end up getting the pong from the full node too late
+      // in that case the websocket would be closed but we know the connection is not down because we are receiving
+      // other messages. Because of that we just reset the timeoutTimer when we receive a message that is not a pong
+      if (this.timeoutTimer) {
+        clearTimeout(this.timeoutTimer);
+        this.timeoutTimer = setTimeout(() => this.onConnectionDown(), this.connectionTimeout);
+      }
+    }
+    this.emit(_type, message)
+  }
+
+  /**
+   * Ping method to check if server is still alive
+   */
+  sendPing() {
```

Comment:
It doesn't feel right to me, maybe in another implementation of a `WebSocket` class we would just send a string `PING` instead of an object

But I can add the `key` parameter if you think we should

Reply:
I think I agree with you, this doesn't seem right. The only thing that bothers me is that any change in this method if we find a bug will need to be replicated in all subclasses and this is even worse to me.

We could also have a new abstract method `getPingMessage`, which is the only different method in both classes, the `sendPing` would be unique and like this:

```
sendPing() {
  if (this.latestPingDate) {
      // Skipping sendPing. Still waiting for pong...
      return;
    }
    const msg = this.getPingMessage();
    this.latestPingDate = new Date();
    this.timeoutTimer = setTimeout(() => this.onConnectionDown(), this.connectionTimeout);
    this.sendMessage(msg);
  }
```

What do you think? I'm okay with whatever you choose.

Reply:
I like it, done in f246cdd5911f141d5b344a89ba6389f0dfc7c07d